lib/fluent/plugin/buffer.rb

```diff
@@ -283,7 +283,13 @@ def write(metadata_and_data, format: nil, size: nil, enqueue: false)
               chunk.mon_enter # add lock to prevent to be committed/rollbacked from other threads
               operated_chunks << chunk
               if chunk.staged?
-                staged_bytesize += adding_bytesize
+                #
+                # https://github.com/fluent/fluentd/issues/2712
+                # write_once is supposed to write to a chunk only once
+                # but this block **may** run multiple times from write_step_by_step and previous write may be rollbacked
+                # So we should be counting the stage_size only for the last successful write
+                #
+                staged_bytesizes_by_chunk[chunk] = adding_bytesize
```

Comment:
I think the following situation can happen. what do you think of this?

1. `staged_bytesizes_by_chunk` stores chunk1 
2. rollback happens in write_step_by_step (ShouldRetry raises)
3. another thread euqueues chunk1 before this thread enters into https://github.com/fluent/fluentd/blob/065751f5c370b70ba8ab5e12aeced0c4a6138eb8/lib/fluent/plugin/buffer.rb#L673
4. create new chunk(chunk2) in https://github.com/fluent/fluentd/blob/065751f5c370b70ba8ab5e12aeced0c4a6138eb8/lib/fluent/plugin/buffer.rb#L663
5. `staged_bytesizes_by_chunk` stores chunk2 (still remtains chunk1 in `staged_bytesizes_by_chunk`)

Reply:
@ganmacs Steps (3) can't happen because `enqueue_chunk` can't get a lock and waits until the lock is released by the thread which is writing to the chunk.

The lock is acquired here: https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/buffer.rb#L279
And released here:
https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/buffer.rb#L319

So `enqueue_chunk` can happen only after the commit is done. 
Hope this helps.

Reply:
Make sense. You're right. other thread can get the chunk but can't enter https://github.com/fluent/fluentd/blob/065751f5c370b70ba8ab5e12aeced0c4a6138eb8/lib/fluent/plugin/buffer.rb#L674 .